tests/cpp/c_codegen_test.cc

```diff
@@ -0,0 +1,127 @@
+ * under the License.
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <tvm/driver/driver_api.h>
+#include <tvm/ir/type.h>
+#include <tvm/node/reflection.h>
+#include <tvm/runtime/metadata.h>
+#include <tvm/runtime/module.h>
+#include <tvm/target/target.h>
+#include <tvm/te/operation.h>
+
+TEST(CCodegen, MainFunctionOrder) {
+  using namespace tvm;
+  using namespace tvm::te;
+
+  std::string tvm_module_main = std::string(runtime::symbol::tvm_module_main);
+
+  tvm::Target target_c = tvm::Target("c -keys=cpu -link-params=0");
+
+  const int n = 4;
+  Array<PrimExpr> shape{n};
+
+  auto A = placeholder(shape, DataType::Float(32), "A");
+  auto B = placeholder(shape, DataType::Float(32), "B");
+
+  auto elemwise_add = compute(
+      A->shape, [&A, &B](PrimExpr i) { return A[i] + B[i]; }, "elemwise_add");
+
+  auto fcreate = [=]() {
+    With<Target> llvm_scope(target_c);
+    return create_schedule({elemwise_add->op});
+  };
+
+  auto args = Array<Tensor>({A, B, elemwise_add});
+
+  std::unordered_map<Tensor, Buffer> binds;
+  auto lowered = LowerSchedule(fcreate(), args, "elemwise_add", binds);
+  Map<tvm::Target, IRModule> inputs = {{target_c, lowered}};
+  runtime::Module module = build(inputs, Target());
+  Array<String> functions = module->GetFunction("get_func_names", false)();
+
+  ICHECK(functions.back().compare(tvm_module_main) == 0);
+}
+
+auto BuildLowered(std::string op_name, tvm::Target target) {
+  using namespace tvm;
+  using namespace tvm::te;
+
+  // The shape of input tensors.
+  const int n = 4;
+  Array<PrimExpr> shape{n};
+
+  auto A = placeholder(shape, DataType::Float(32), "A");
+  auto B = placeholder(shape, DataType::Float(32), "B");
+
+  auto op = compute(
+      A->shape, [&A, &B](PrimExpr i) { return A[i] + B[i]; }, op_name);
+
+  auto fcreate_s = [=]() {
+    With<Target> llvm_scope(target);
+    return create_schedule({op->op});
+  };
+
+  auto args = Array<Tensor>({A, B, op});
+  std::unordered_map<Tensor, Buffer> binds;
+  auto lowered_s = LowerSchedule(fcreate_s(), args, op_name, binds);
+  return lowered_s;
+}
+
+bool IsSorted(tvm::Map<tvm::Target, tvm::IRModule> inputs) {
+  std::vector<std::string> schedule_names;
+  for (auto const& module : inputs) {
+    for (auto const& func : module.second->functions) {
+      schedule_names.push_back(func.first->name_hint);
+    }
+  }
+  return std::is_sorted(schedule_names.begin(), schedule_names.end());
+}
+
+TEST(CCodegen, FunctionOrder) {
+  using testing::_;
+  using testing::ElementsAre;
+  using testing::StrEq;
+  using namespace tvm;
+  using namespace tvm::te;
+
+  Target target = Target("c -keys=cpu -link-params=0");
+
+  // add schedules in reverse order
+  Map<tvm::Target, IRModule> inputs;
+  inputs.Set(Target("c -keys=cpu -link-params=0"), BuildLowered("op_2", target));
+  inputs.Set(Target("c -keys=cpu -link-params=0"), BuildLowered("op_1", target));
+
+  uint32_t counter = 9;
+  while (IsSorted(inputs)) {
```

Comment:
could also write as a for-loop:
```suggestion
  for (uint32_t counter = 99; !IsSorted(inputs) && counter > 0; counter--)
```

Reply:
added.